core/src/databases/database.rs

```diff
@@ -58,16 +58,83 @@ impl Database {
         match self.db_type {
             DatabaseType::REMOTE => Err(anyhow!("Remote DB not implemented.")),
             DatabaseType::LOCAL => {
-                let rows = self.get_rows(store).await?;
+                // If there is a schema cache, return it.
+                if let Some(schema) = self.get_schema_from_cache(store.clone()).await? {
+                    return Ok(schema);
+                }
 
-                rows.par_iter()
+                // Compute & cache all table schemas from scratch.
+                // Kept for backward compatibility.
+                // TODO(@fontanierh): remove this once all databases have a schema cache.
+
+                // Fetch the rows to infer the schema.
+                let rows = self.get_rows(store.clone()).await?;
+                let schema_tables = rows
+                    .par_iter()
                     .map(|(table, r)| {
                         Ok(DatabaseSchemaTable::new(
                             table.clone(),
-                            TableSchema::from_rows(&r)?,
+                            TableSchema::from_rows(r)?,
                         ))
                     })
-                    .collect::<Result<Vec<_>>>()
+                    .collect::<Result<Vec<_>>>()?;
+
+                // Concurrently update the schema caches.
+                futures::future::try_join_all(schema_tables.iter().map(|st| {
+                    store.update_database_table_schema(
+                        &self.project,
+                        &self.data_source_id,
+                        &self.database_id,
+                        st.table().table_id(),
+                        &st.schema,
+                    )
+                }))
+                .await?;
+
+                Ok(schema_tables)
+            }
+        }
+    }
+
+    async fn get_schema_from_cache(
+        &self,
+        store: Box<dyn Store + Sync + Send>,
+    ) -> Result<Option<Vec<DatabaseSchemaTable>>> {
+        match self.db_type {
+            DatabaseType::REMOTE => Err(anyhow!("Remote DB not implemented.")),
+            DatabaseType::LOCAL => {
+                // If there is no schema cache, then we are loading all the tables twice (once here
+                // and once in get_rows).
+                // Once all databases have a schema cache, we can remove this function and it won't be an issue.
+                // TODO(@fontanierh): remove this once all databases have a schema cache.
+                let (tables, _) = store
+                    .list_databases_tables(
+                        &self.project,
+                        &self.data_source_id,
+                        &self.database_id,
+                        None,
+                    )
+                    .await?;
+
+                // If some table don't have a schema cache, return None.
+                if !tables.iter().all(|t| t.schema_cache().is_some()) {
+                    return Ok(None);
```

Comment:
@spolu Right now if a DB has at least one empty table, we'll return None and we'll end up computing the schema from scratch.

Once we're done migrating (that is once I call `get_schema` once on the few DBs we have in prod) we can simply remove the condition and just remove the empty tables from the output schema.